Added more LSP support for Python by adding ruff-lsp support.

frejanordsiek opened the pull request:
As it says on the tin, adds support for ruff-lsp for Python (https://github.com/charliermarsh/ruff-lsp) including adding it to the documentation (not completely sure I got it right).

ruff-lsp was brought up in Issue #3876 where others have been working on it and had some success in getting it working. 

This implementation provides customization options for every setting ruff-lsp has but one (showNotification since I am not sure what all values it supports). It is heavily based off of clients/lsp-pylsp.el, with some ideas from the other ones. Note, unlike the version from Nikolai Prokoschenko (@rassie) in Issue #3876, this does not have poetry support.

I set its priority to -2, putting it behind pylsp which most people seem to use and prefer (no surprises for users).

yyoncho commented on clients/lsp-ruff-lsp.el:
```diff
@@ -0,0 +1,95 @@
+                :path lsp-ruff-lsp-ruff-path
+                :interpreter (vector lsp-ruff-lsp-python-path)
+                :organizeImports (if lsp-ruff-lsp-advertize-organize-imports t :json-false)
+                :fixAll (if lsp-ruff-lsp-advertize-fix-all t :json-false))))))
```
nit: lsp-json-bool

frejanordsiek replied:
Didn't realize that existed (not a surprise). Thank you. I have changed the code to use it.

yyoncho commented on clients/lsp-ruff-lsp.el:
```diff
@@ -0,0 +1,95 @@
+  :group 'lsp-ruff-lsp)
+
+
+(lsp-register-client
```
shouldn't that be add-on? = t?

frejanordsiek replied:
I don't know. Reading the FAQ, it seems like ``add-on?`` being ``t`` means load in parallel to other LSPs at the same priority? Is compatibility something to worry about? Only a few of the client modules use it:

* lsp-eslint.el
* lsp-emmet.el
* lsp-sorbet.el
* lsp-angular.el
* lsp-camel.el

None of the python ones do at present (though, of course, that doesn't mean that they shouldn't).

So basically, I don't know enough about the option to know.

rassie replied:
Ruff doesn't provide anything you'd like a language server to have, e.g. code navigation, completion etc., it's just a linter. So running it standalone is usually not what you want, but pairing it with e.g. pylsp is extremely natural -- you'd have two servers running, both providing code actions and diagnostics at the same time (e.g. pylsp would provide mypy diagnostics, while ruff would be linting). It's exactly what we want and need (like I said, it's basically the same situation as with eslint). We haven't seen addon LSPs with Python yet, since pylint and pyflakes and flake8 and all the other linters have been typically used as plugins for pylsp, but ruff is mostly used standalone (but there is a plugin for pylsp too).

frejanordsiek replied:
So that is why the other python LSPs have not done it. Makes sense. Speaking of, the option to enable ruff for pylsp does need to have its customization options added but that is for another day.

Thank you for the explanations.

frejanordsiek commented:
I rebased the PR, added the ``showNotifications`` option, added the missing values for the ``logLevel`` option, and added the missing ``importStrategy`` option (it is declared in the ruff-lsp code as well as its vscode plugin, but not in ruff-lsp's README).

frejanordsiek commented:
I do not know why the check on macos-latest with emacs 26.3 is failing

nasyxx commented:
Any update for this?  Will it merge to the main branch soon?

yyoncho commented:
@frejanordsiek thank you!
